std/node/_errors.ts

```diff
@@ -20,33 +20,101 @@
 // DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 // OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
 // USE OR OTHER DEALINGS IN THE SOFTWARE.
+/************ NOT IMPLEMENTED
+* ERR_INVALID_ARG_VALUE
+* ERR_INVALID_MODULE_SPECIFIER
+* ERR_INVALID_PACKAGE_TARGET
+* ERR_INVALID_URL_SCHEME
+* ERR_MANIFEST_ASSERT_INTEGRITY
+* ERR_MISSING_ARGS
+* ERR_MODULE_NOT_FOUND
+* ERR_PACKAGE_PATH_NOT_EXPORTED
+* ERR_QUICSESSION_VERSION_NEGOTIATION
+* ERR_REQUIRE_ESM
+* ERR_SOCKET_BAD_PORT
+* ERR_TLS_CERT_ALTNAME_INVALID
+* ERR_UNHANDLED_ERROR
+* ERR_WORKER_INVALID_EXEC_ARGV
+* ERR_WORKER_PATH
+* ERR_QUIC_ERROR
+* ERR_SOCKET_BUFFER_SIZE //System error, shouldn't ever happen inside Deno
+* ERR_SYSTEM_ERROR //System error, shouldn't ever happen inside Deno
+* ERR_TTY_INIT_FAILED //System error, shouldn't ever happen inside Deno
+* ERR_INVALID_PACKAGE_CONFIG // package.json stuff, probably useless
+*************/
 
 import { unreachable } from "../testing/asserts.ts";
 
-// It will do so until we'll have Node errors completely ported (#5944):
+/**
+ * All error instances in Node have additional methods and properties
+ * This export class is meant to be extended by these instances abstracting native JS error instances
+ */
+export class NodeErrorAbstraction extends Error {
+  code: string;
 
-// Ref: https://github.com/nodejs/node/blob/50d28d4b3a616b04537feff014aa70437f064e30/lib/internal/errors.js#L251
-// Ref: https://github.com/nodejs/node/blob/50d28d4b3a616b04537feff014aa70437f064e30/lib/internal/errors.js#L299
-// Ref: https://github.com/nodejs/node/blob/50d28d4b3a616b04537feff014aa70437f064e30/lib/internal/errors.js#L325
-// Ref: https://github.com/nodejs/node/blob/50d28d4b3a616b04537feff014aa70437f064e30/lib/internal/errors.js#L943
-class ERR_INVALID_ARG_TYPE extends TypeError {
-  code = "ERR_INVALID_ARG_TYPE";
+  constructor(name: string, code: string, message: string) {
+    super(message);
+    this.code = code;
+    this.name = name;
+    //This number changes dependending on the name of this class
+    //20 characters as of now
+    this.stack = this.stack && `${name} [${this.code}]${this.stack.slice(20)}`;
+  }
+
+  toString() {
+    return `${this.name} [${this.code}]: ${this.message}`;
+  }
+}
+
+export class NodeError extends NodeErrorAbstraction {
+  constructor(code: string, message: string) {
+    super(Error.prototype.name, code, message);
+  }
+}
```

Comment:
An alternative pattern here, could be the Node.js error helper pattern, which would make this file less verbose:

```js
function E(code, msg, base = Error) {
  const err = class extends base {
    code = code;
 
    constructor() {
      super(typeof msg === "string" ? msg : msg(...arguments));
      const { name } = this;
      // Add the error code to the name to include it in the stack trace.
      this.name = `${name} [${this.code}]`;
      // Access the stack to generate the error message including the error code from the name.
      this.stack;
      // Reset the name to the actual name.
      this.name = name;
    }
  };
  return err;
}
```

in addition this would avoid the need to have a `NodeRangeError` etc for all base error types.

Reply:
I opted for the class based approach cause it's visually simpler and allows us to use the import/export system. In my head I saw it this way: If I were to produce a bundle from std/node, would I prefer to export an object that contains every single error in Node? Or the exact error classes that I'm needing for my app?

Reply:
That's a good point, and definitely a property worth ensuring in the output.

Note that the purity of the `E` function above is determinable by RollupJS:

http://rollupjs.org/repl/?version=2.33.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMEVSUl9TT01FX0VSUiUyMCU3RCUyMGZyb20lMjAnLiUyRmVycm9ycy5qcyclM0IlNUNudGhyb3clMjBuZXclMjBFUlJfU09NRV9FUlIoKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmVycm9ycy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJmdW5jdGlvbiUyMEUoY29kZSUyQyUyMG1zZyUyQyUyMGJhc2UlMjAlM0QlMjBFcnJvciklMjAlN0IlNUNuJTIwJTIwY29uc3QlMjBlcnIlMjAlM0QlMjBjbGFzcyUyMGV4dGVuZHMlMjBiYXNlJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGNvZGUlMjAlM0QlMjBjb2RlJTNCJTVDbiUyMCU1Q24lMjAlMjAlMjAlMjBjb25zdHJ1Y3RvcigpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMHN1cGVyKHR5cGVvZiUyMG1zZyUyMCUzRCUzRCUzRCUyMCU1QyUyMnN0cmluZyU1QyUyMiUyMCUzRiUyMG1zZyUyMCUzQSUyMG1zZyguLi5hcmd1bWVudHMpKSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjBjb25zdCUyMCU3QiUyMG5hbWUlMjAlN0QlMjAlM0QlMjB0aGlzJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyRiUyRiUyMEFkZCUyMHRoZSUyMGVycm9yJTIwY29kZSUyMHRvJTIwdGhlJTIwbmFtZSUyMHRvJTIwaW5jbHVkZSUyMGl0JTIwaW4lMjB0aGUlMjBzdGFjayUyMHRyYWNlLiU1Q24lMjAlMjAlMjAlMjAlMjAlMjB0aGlzLm5hbWUlMjAlM0QlMjAlNjAlMjQlN0JuYW1lJTdEJTIwJTVCJTI0JTdCdGhpcy5jb2RlJTdEJTVEJTYwJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyRiUyRiUyMEFjY2VzcyUyMHRoZSUyMHN0YWNrJTIwdG8lMjBnZW5lcmF0ZSUyMHRoZSUyMGVycm9yJTIwbWVzc2FnZSUyMGluY2x1ZGluZyUyMHRoZSUyMGVycm9yJTIwY29kZSUyMGZyb20lMjB0aGUlMjBuYW1lLiU1Q24lMjAlMjAlMjAlMjAlMjAlMjB0aGlzLnN0YWNrJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyRiUyRiUyMFJlc2V0JTIwdGhlJTIwbmFtZSUyMHRvJTIwdGhlJTIwYWN0dWFsJTIwbmFtZS4lNUNuJTIwJTIwJTIwJTIwJTIwJTIwdGhpcy5uYW1lJTIwJTNEJTIwbmFtZSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlNUNuJTIwJTIwJTdEJTNCJTVDbiUyMCUyMHJldHVybiUyMGVyciUzQiU1Q24lN0QlNUNuJTVDbmV4cG9ydCUyMGNvbnN0JTIwRVJSX1NPTUVfRVJSJTIwJTNEJTIwRSgnRVJSX1NPTUVfRVJSJyUyQyUyMCdtc2cnKSUzQiU1Q25leHBvcnQlMjBjb25zdCUyMEVSUl9BTk9USEVSX0VSUiUyMCUzRCUyMEUoJ0VSUl9BTk9USEVSX0VSUiclMkMlMjAnbXNnJyklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

Reply:
This solution seems viable and would definitely save a good chunk of the current code. However it would make us lose the type safety we currently have. I know TypeScript is like the CSS of JS and doesn't really matter in the end, but this niceties make a difference in the end.

Reply:
Certainly, it was only a suggestion and not a request.